ql/ql/src/codeql_ql/StructuredLogs.qll

```diff
@@ -31,9 +33,20 @@ class Array extends JSON::Array {
 
   int getNumber(int i) { result = this.getChild(i).(JSON::Number).getValue().toInt() }
 
+  float getFloat(int i) { result = this.getChild(i).(JSON::Number).getValue().toFloat() }
+
   Array getArray(int i) { result = this.getChild(i) }
 }
 
+/**
+ * Gets the i'th non-negative number in `a`.
+ *
+ * This is needed because the evaluator log is padded with -1s in some cases.
```

Comment:
Hmm... The `-1`s correspond to the empty lines in the corresponding RA. We would need to also "rank" those out so that the correspondence is maintained.

What exactly does the `-1` mess up?

Reply:
> Hmm... The `-1`s correspond to the empty lines in the corresponding RA. We would need to also "rank" those out so that the correspondence is maintained.

Yep, that's true. I suppose we should have a class for `RA` that has a `get(int)` predicate which would then do the ranking?

> What exactly does the `-1` mess up?

It's just a weird artifact of the way the structured log is organized, and I don't think it messes up anything crucially.

The only area I'm aware of in DCA is when computing the join order metric for recursive we predicates. In this case, we bucket each iteration into buckets indexed by the pipeline ordering (i.e., `order_50000`, `order_500000`, `standard`, etc.) and sum each bucket's tuple counts to compute the join order metric for each ordering. To avoid repeatedly adding `-1` to itself we then filter out those negative numbers to avoid computing "negative join order metrics".

If you'd rather keep those `-1`s we can just to that filtering in the ql-for-ql query.

Reply:
Nah, that's fine. I don't think there's any useful signal in the blank lines that we couldn't reconstruct otherwise. (I know they indicate the beginning of something that isn't dependent on the preceding line, but we should have this information in a much better format once we have the RA parsing done.)